{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing the RareLink-CDM\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RD CDM v2.0: 'res/RD CDM v2.0.xlsx'\n",
    "\n",
    "\n",
    "RareLink Data Dictionary: '../res/RareLink_v2.0_DataDictionary.csv'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 - Defining the RD CDM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 Resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards.code_system import ORDO, HPO, MONDO, OMIM, LOINC, HGNC, HGVS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards.code_system import CodeSystem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "NCIT = CodeSystem(\n",
    "    name='National Cancer Institute Thesaurus (NCIT)', \n",
    "    namespace_prefix='ncit', \n",
    "    url='https://ncithesaurus.org/'\n",
    ")\n",
    "\n",
    "UO = CodeSystem(\n",
    "    name='Units of Measurement Ontology (UO)', \n",
    "    namespace_prefix='uo', \n",
    "    url='http://purl.obolibrary.org/obo/uo.owl'\n",
    ")\n",
    "\n",
    "NCBITaxon = CodeSystem(\n",
    "    name='NCBI organismal classification', \n",
    "    namespace_prefix='ncbitaxon', \n",
    "    url='https://www.ncbi.nlm.nih.gov/taxonomy'\n",
    ")\n",
    "\n",
    "GENO = CodeSystem(\n",
    "    name='GENO: The Genotype Ontology', \n",
    "    namespace_prefix='geno', \n",
    "    url='http://www.genoontology.org/'\n",
    ")\n",
    "\n",
    "SO = CodeSystem(\n",
    "    name='Sequence Ontology (SO)', \n",
    "    namespace_prefix='so', \n",
    "    url='http://www.sequenceontology.org/'\n",
    ")\n",
    "\n",
    "ICD10CM = CodeSystem(\n",
    "    name='ICD-10-CM', \n",
    "    namespace_prefix='icd10cm', \n",
    "    url='https://www.cdc.gov/nchs/icd/icd10cm.htm'\n",
    ")\n",
    "\n",
    "SNOMED = CodeSystem(\n",
    "    name='SNOMED CT', \n",
    "    namespace_prefix='snomed', \n",
    "    url='https://www.snomed.org/snomed-ct'\n",
    ")\n",
    "\n",
    "ICD11 = CodeSystem(\n",
    "    name='ICD-11', \n",
    "    namespace_prefix='icd11', \n",
    "    url='https://icd.who.int/en'\n",
    ")\n",
    "\n",
    "HL7FHIR = CodeSystem(\n",
    "    name='HL7 FHIR', \n",
    "    namespace_prefix='hl7fhir', \n",
    "    url='https://www.hl7.org/fhir/'\n",
    ")\n",
    "\n",
    "GA4GH = CodeSystem(\n",
    "    name='Global Alliance for Genomics and Health - Phenopacket Schema', \n",
    "    namespace_prefix='ga4gh', \n",
    "    url='https://www.ga4gh.org/'\n",
    ")\n",
    "\n",
    "ISO3166 = CodeSystem(\n",
    "    name='ISO 3166-1:2020(en) alpha-2 and alpha-3 country codes', \n",
    "    namespace_prefix='iso3166', \n",
    "    url='https://www.iso.org/iso-3166-country-codes.html'\n",
    ")\n",
    "\n",
    "ICF = CodeSystem(\n",
    "    name='International Classification of Functioning, Disability and Health (ICF)', \n",
    "    namespace_prefix='icf', \n",
    "    url='https://www.who.int/classifications/icf/en/'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "resources = [\n",
    "    HPO.set_version(\"2024-08-13\"),\n",
    "    MONDO.set_version(\"2024-09-03\"),\n",
    "    OMIM.set_version(\"2024-09-12\"),\n",
    "    ORDO.set_version(\"2024-09-12\"),\n",
    "    NCBITaxon.set_version(\"2024-07-03\"),\n",
    "    LOINC.set_version(\"2.78\"),\n",
    "    HGNC.set_version(\"2024-08-23\"),\n",
    "    GENO.set_version(\"2023-10-08\"),\n",
    "    NCIT.set_version(\"24.04e\"),\n",
    "    SO.set_version(\"2.6\"),\n",
    "    ICD10CM.set_version(\"2024-09-01\"),\n",
    "    SNOMED.set_version(\"2024-09-01\"),\n",
    "    ICD11.set_version(\"2024-09-01\"),\n",
    "    HL7FHIR.set_version(\"v4.0.1\"),\n",
    "    GA4GH.set_version(\"v2.0\"),\n",
    "    ISO3166.set_version(\"2020(en)\"),\n",
    "    HGVS.set_version(\"2024-09-01\"),\n",
    "    ICF.set_version(\"2024-09-01\")\n",
    "]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.2 Fields of the data model and their value sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards.value_set import ValueSet\n",
    "from phenopacket_mapper.data_standards import Date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Formal Criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "vs_1_1 = ValueSet(\n",
    "    elements=[str],\n",
    "    name=\"Value set for 1.1 Pseudonym\",\n",
    "    description=\"The (local) patient-related Identification code.\"\n",
    ")\n",
    "\n",
    "vs_1_2 = ValueSet(\n",
    "    elements=[Date],\n",
    "    name=\"Value set for 1.2 Date of Admission\",\n",
    "    description=\"The date of admission or data capture of the individual.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Personal Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards.code import Coding\n",
    "\n",
    "\n",
    "vs_2_1 = ValueSet(\n",
    "    elements=[Date],  # Keep non-string elements as is\n",
    "    name=\"Value set for 2.1 Date of Birth\",\n",
    "    description=(\"The individual's date of birth. If the exact month or day \"\n",
    "                 \"is allowed to be captured or not known, select the 1st day \"\n",
    "                 \"of the month or the 1st month of the year, respectively.\")\n",
    ")\n",
    "\n",
    "vs_2_2 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"248152002\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"248153007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"184115007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"32570691000036108\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"1220561009\")\n",
    "    ],\n",
    "    name=\"Value set for 2.2 Sex at Birth\",\n",
    "    description=(\"The individual's sex that was assigned at birth.\")\n",
    ")\n",
    "\n",
    "vs_2_3 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"734875008\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"734876009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"80427008\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"65162001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"35111009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"403760006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"78317008\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"10567003\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"48930007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"74964007\")\n",
    "    ],\n",
    "    name=\"Value set for 2.3 Karyotypic Sex\",\n",
    "    description=(\"The chromosomal sex of an individual.\")\n",
    ")\n",
    "\n",
    "vs_2_4 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"446141000124107\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"446151000124109\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"394743007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"33791000087105\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"1220561009\")\n",
    "    ],\n",
    "    name=\"Value set for 2.4 Gender Identity\",\n",
    "    description=(\"The self-assigned gender of the individual.\")\n",
    ")\n",
    "\n",
    "\n",
    "# 2.5 Country of Birth\n",
    "vs_2_5 = ValueSet(\n",
    "    elements=[ISO3166],\n",
    "    name=\"Value set for 2.5 Country of Birth\",\n",
    "    description=\"The country in which the individual was born.\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Formal Criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.1 Vital Status\n",
    "vs_3_1 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"438949009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"419099009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"399307001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"185924006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")\n",
    "    ],\n",
    "    name=\"Value set for 3.1 Vital Status\",\n",
    "    description=(\"The individual’s general clinical status or vital status.\")\n",
    ")\n",
    "\n",
    "# 3.2 Time of Death\n",
    "vs_3_2 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 3.2 Time of Death\",\n",
    "    description=(\"If deceased, the individual’s date of death. If the specific \"\n",
    "                 \"month or day is not known, select the 1st day of the \"\n",
    "                 \"month or the 1st month of the year, respectively.\")\n",
    ")\n",
    "\n",
    "# 3.3 Cause of Death [ICD10CM]\n",
    "vs_3_3 = ValueSet(\n",
    "    elements=[ICD10CM],  # Based on ICD10CM coding\n",
    "    name=\"Value set for 3.3 Cause of Death [ICD10CM]\",\n",
    "    description=(\"If deceased, the individual’s primary cause of death \"\n",
    "                 \"(i.e. according to the death certificate).\")\n",
    ")\n",
    "\n",
    "# 3.4 Age Category\n",
    "vs_3_4 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"3658006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"713153009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"255398004\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"263659003\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"41847000\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"303112003\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"419099009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")\n",
    "    ],\n",
    "    name=\"Value set for 3.4 Age Category\",\n",
    "    description=(\"The individual's age category at the time of data capture (1.2).\")\n",
    ")\n",
    "\n",
    "# 3.5 Length of Gestation at Birth [weeks+days]\n",
    "vs_3_5 = ValueSet(\n",
    "    elements=[str],\n",
    "    name=\"Value set for 3.5 Length of Gestation at Birth [weeks+days]\",\n",
    "    description=(\"The length of gestation at birth in weeks and days.\")\n",
    ")\n",
    "\n",
    "# 3.6 Undiagnosed RD Case\n",
    "vs_3_6 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373066001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373067005\")\n",
    "    ],\n",
    "    name=\"Value set for 3.6 Undiagnosed RD Case\",\n",
    "    description=(\"Identifies cases where an RD diagnosis has not been established.\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Care Pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.1 Encounter Start\n",
    "vs_4_1 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 4.1 Encounter Start\",\n",
    "    description=(\"The date of the start of the encounter.\")\n",
    ")\n",
    "\n",
    "# 4.2 Encounter End\n",
    "vs_4_2 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 4.2 Encounter End\",\n",
    "    description=(\"The date of the end of the encounter.\")\n",
    ")\n",
    "\n",
    "# 4.3 Encounter Status\n",
    "vs_4_3 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"planned\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"arrived\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"triaged\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"in-progress\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"onleave\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"finished\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"cancelled\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"entered-in-error\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"unknown\")\n",
    "    ],\n",
    "    name=\"Value set for 4.3 Encounter Status\",\n",
    "    description=(\"The status of an encounter of the individual at the time of \"\n",
    "                 \"data capture.\")\n",
    ")\n",
    "\n",
    "# 4.4 Encounter Class\n",
    "vs_4_4 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"amb\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"imp\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"obsenc\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"emer\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"vr\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"hh\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"rdc\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")\n",
    "    ],\n",
    "    name=\"Value set for 4.4 Encounter Class\",\n",
    "    description=(\"The class of an encounter of the individual at the time of \"\n",
    "                 \"data capture.\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Disease\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "vs_5_1 = ValueSet(\n",
    "    elements=[MONDO, ORDO, ICD10CM, ICD11, OMIM],\n",
    "    name=\"Value set for 5.1 Disease [MONDO, ORDO, ICD10CM, ICD11, OMIM]\",\n",
    "    description=(\"A disease that the individual was affected by. If a genetic \"\n",
    "                 \"diagnosis or subtypes were diagnosed, please also provide the\"\n",
    "                 \"respective OMIM_g and OMIM_p codes.\")\n",
    ")\n",
    "\n",
    "# 5.2 Verification Status\n",
    "vs_5_2 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"unconfirmed\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"provisional\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"differential\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"confirmed\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"refuted\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"entered-in-error\")\n",
    "    ],\n",
    "    name=\"Value set for 5.2 Verification Status\",\n",
    "    description=(\"The verification status of the disease.\")\n",
    ")\n",
    "\n",
    "# 5.3 Age at Onset\n",
    "vs_5_3 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"118189007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"3950001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"410672004\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")\n",
    "    ],\n",
    "    name=\"Value set for 5.3 Age at Onset\",\n",
    "    description=(\"The age at the onset of the first symptoms or signs\"\n",
    "                 \"of the disease.\")\n",
    ")\n",
    "\n",
    "# 5.4 Date of Onset\n",
    "vs_5_4 = ValueSet(\n",
    "    elements=[Date],\n",
    "    name=\"Value set for 5.4 Date of Onset\",\n",
    "    description=(\"The date at onset of first symptoms or signs of the disease.\"\n",
    "                 \"If the specific month or day is not known, select the 1st day\"\n",
    "                  \"of the month or the 1st month of the year, respectively.\")\n",
    ")\n",
    "\n",
    "# 5.5 Age at Diagnosis\n",
    "vs_5_5 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"118189007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"3950001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"410672004\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")\n",
    "    ],\n",
    "    name=\"Value set for 5.5 Age at Diagnosis\",\n",
    "    description=(\"The individual’s age when the diagnosis was made.\")\n",
    ")\n",
    "\n",
    "# 5.6 Date of Diagnosis\n",
    "vs_5_6 = ValueSet(\n",
    "    elements=[Date],\n",
    "    name=\"Value set for 5.6 Date of Diagnosis\",\n",
    "    description=(\"If the specific month or day is not known, select the 1st\"\n",
    "                 \" day of the month or the 1st month \"\n",
    "                 \"of the year, respectively.\")\n",
    ")\n",
    "\n",
    "# 5.7 Body Site [SNOMED CT]\n",
    "vs_5_7 = ValueSet(\n",
    "    elements=[SNOMED],\n",
    "    name=\"Value set for 5.7 Body Site [SNOMED CT]\",\n",
    "    description=(\"The specific body site affected by disease is encoded using all \"\n",
    "                 \"descendants of SCT Body Structure (123037004).\")\n",
    ")\n",
    "\n",
    "# 5.8 Clinical Status\n",
    "vs_5_8 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"active\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"recurrence\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"relapse\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"inactive\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"remission\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"resolved\")\n",
    "    ],\n",
    "    name=\"Value set for 5.8 Clinical Status\",\n",
    "    description=(\"The clinical status of the disease indicates whether\"\n",
    "                  \"it is active, inactive, or resolved.\")\n",
    ")\n",
    "\n",
    "# 5.9 Severity\n",
    "vs_5_9 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"24484000\"),  # Severe\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"6736007\"),   # Moderate\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"255604002\")  # Mild\n",
    "    ],\n",
    "    name=\"Value set for 5.9 Severity\",\n",
    "    description=(\"The severity of the disease is categorised by clinical evaluation.\")\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Genetic Finddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6.1.1 Genomic Diagnosis [MONDO, OMIM]\n",
    "vs_6_1_1 = ValueSet(\n",
    "    elements=[MONDO, OMIM],\n",
    "    name=\"Value set for 6.1.1 Genomic Diagnosis [MONDO]\",\n",
    "    description=(\"The genetic finding of a variant can be linked to a disease \"\n",
    "                 \"in (5.1) if the same MONDO code is used.\")\n",
    ")\n",
    "\n",
    "# 6.1.2 Progress Status of Interpretation\n",
    "vs_6_1_2 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"unknown_progress\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"in_progress\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"completed\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"solved\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"unsolved\")\n",
    "    ],\n",
    "    name=\"Value set for 6.1.2 Progress Status of Interpretation\",\n",
    "    description=(\"The interpretation has a ProgressStatus that refers to the \"\n",
    "                 \"status of the attempted diagnosis.\")\n",
    ")\n",
    "\n",
    "# 6.1.3 Interpretation Status\n",
    "vs_6_1_3 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"unknown_status\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"rejected\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"candidate\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"contributory\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"causative\")\n",
    "    ],\n",
    "    name=\"Value set for 6.1.3 Interpretation Status\",\n",
    "    description=(\"An enumeration that describes the conclusion made about \"\n",
    "                 \"the genomic interpretation.\")\n",
    ")\n",
    "\n",
    "# 6.1.4 Structural Variant Analysis Method\n",
    "vs_6_1_4 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26406-1\"),  # Karyotyping\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26404-6\"),  # FISH\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26418-6\"),  # PCR\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26419-4\"),  # qPCR (real-time PCR)\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26400-4\"),  # SNP array\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26813-8\"),  # Restriction fragment length polymorphism (RFLP)\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26810-4\"),  # DNA hybridization\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26398-0\"),  # Sequencing\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26415-2\"),  # MLPA\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la46-8\")      # Other\n",
    "    ],\n",
    "    name=\"Value set for 6.1.4 Structural Variant Analysis Method\",\n",
    "    description=(\"The method used to analyse structural variants in the genome.\")\n",
    ")\n",
    "\n",
    "# 6.1.5 Reference Genome\n",
    "vs_6_1_5 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la14032-9\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la14029-5\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la14030-3\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la14031-1\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26806-2\")\n",
    "    ],\n",
    "    name=\"Value set for 6.1.5 Reference Genome\",\n",
    "    description=(\"The reference genome used for analyzing the genetic variant.\")\n",
    ")\n",
    "\n",
    "# 6.1.6 Genetic Mutation String\n",
    "vs_6_1_6 = ValueSet(\n",
    "    elements=[str],  # HGVS strings for genetic mutations\n",
    "    name=\"Value set for 6.1.6 Genetic Mutation String\",\n",
    "    description=(\"An unvalidated (HGVS) string that describes the variant change.\")\n",
    ")\n",
    "\n",
    "# 6.1.7 Genomic DNA Change [g.HGVS]\n",
    "vs_6_1_7 = ValueSet(\n",
    "    elements=[HGVS],  # HGVS strings for genomic DNA change\n",
    "    name=\"Value set for 6.1.7 Genomic DNA Change [g.HGVS]\",\n",
    "    description=(\"The specific change in the genomic DNA sequence encoded \"\n",
    "                 \"with a validated g.HGVS expression.\")\n",
    ")\n",
    "\n",
    "# 6.1.8 Sequence DNA Change [c.HGVS]\n",
    "vs_6_1_8 = ValueSet(\n",
    "    elements=[HGVS],  # HGVS strings for DNA sequence changes\n",
    "    name=\"Value set for 6.1.8 Sequence DNA Change [c.HGVS]\",\n",
    "    description=(\"The specific change in the DNA sequence at the nucleotide level \"\n",
    "                 \"with a validated c.HGVS expression.\")\n",
    ")\n",
    "\n",
    "# 6.1.9 Amino Acid Change [p.HGVS]\n",
    "vs_6_1_9 = ValueSet(\n",
    "    elements=[HGVS],  # HGVS strings for amino acid changes\n",
    "    name=\"Value set for 6.1.9 Amino Acid Change [p.HGVS]\",\n",
    "    description=(\"The specific change in the amino acid sequence resulting from a \"\n",
    "                 \"genetic variant as a validated p.HGVS expression.\")\n",
    ")\n",
    "\n",
    "# 6.1.10 Gene [HGNC-NR]\n",
    "vs_6_1_10 = ValueSet(\n",
    "    elements=[HGNC],  # Gene codes\n",
    "    name=\"Value set for 6.1.10 Gene [HGNC-NR]\",\n",
    "    description=(\"The specific gene or genes that were analyzed or identified in \"\n",
    "                 \"the study.\")\n",
    ")\n",
    "\n",
    "# 6.1.11 Zygosity\n",
    "vs_6_1_11 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6705-3\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6706-1\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26217-2\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26220-6\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6707-9\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6703-8\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6704-6\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"53034-5_other\")\n",
    "    ],\n",
    "    name=\"Value set for 6.1.11 Zygosity\",\n",
    "    description=(\"The zygosity of the genetic variant.\")\n",
    ")\n",
    "\n",
    "# 6.1.12 Genomic Source Class\n",
    "vs_6_1_12 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6683-2\"),  # Germline\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6684-0\"),  # Somatic\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la10429-1\"), # Fetal\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la18194-3\"), # Likely germline\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la18195-0\"), # Likely somatic\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la18196-8\"), # Likely fetal\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la18197-6\"), # Unknown genomic origin\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26807-0\")  # De novo\n",
    "    ],\n",
    "    name=\"Value set for 6.1.12 Genomic Source Class\",\n",
    "    description=(\"The classification of the genomic source, such as germline, \"\n",
    "                 \"somatic, or other origins.\")\n",
    ")\n",
    "\n",
    "# 6.1.13 DNA Change Type\n",
    "vs_6_1_13 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la9658-1\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6692-3\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6686-5\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6687-3\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6688-1\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6689-9\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6690-7\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"48019-4_other\")\n",
    "    ],\n",
    "    name=\"Value set for 6.1.13 DNA Change Type\",\n",
    "    description=(\"Thevariant’s type of DNA change, such as point mutation,\"\n",
    "                 \"deletion, insertion, or other types.\")\n",
    ")\n",
    "\n",
    "vs_6_1_13a = ValueSet(\n",
    "    elements=[LOINC],\n",
    "    name=\"Other 6.1.13 DNA Change Type in LOINC\",\n",
    "    description=\"Value set for other DNA change types in LOINC.\"\n",
    ")\n",
    "\n",
    "# 6.1.14 Clinical Significance [ACMG]\n",
    "vs_6_1_14 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6668-3\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26332-9\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26333-7\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la26334-5\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la6675-8\"),\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la4489-6\")\n",
    "    ],\n",
    "    name=\"Value set for 6.1.14 Clinical Significance [ACMG]\",\n",
    "    description=(\"The clinical significance of the genetic variant, indicating its \"\n",
    "                 \"impact on health and disease.\")\n",
    ")\n",
    "\n",
    "# 6.1.15 Therapeutic Actionability\n",
    "vs_6_1_15 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"unknown_actionability\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"not_actionable\"),\n",
    "        Coding(system=CodeSystem(name=\"GA4GH\", namespace_prefix=\"ga4gh\"),\n",
    "               code=\"actionable\")\n",
    "    ],\n",
    "    name=\"Value set for 6.1.15 Therapeutic Actionability\",\n",
    "    description=(\"This field flags the variant as being a candidate for treatment/clinical \"\n",
    "                 \"intervention, which could improve the clinical outcome.\")\n",
    ")\n",
    "\n",
    "# 6.1.16 Clinical Annotation Level Of Evidence\n",
    "vs_6_1_16 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la30200-2\"),  # Very strong evidence pathogenic\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la30201-0\"),  # Strong evidence pathogenic\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la30202-8\"),  # Moderate evidence pathogenic\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la30203-6\"),  # Supporting evidence pathogenic\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la30204-4\"),  # Supporting evidence benign\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la30205-1\"),  # Strong evidence benign\n",
    "        Coding(system=CodeSystem(name=\"LOINC\", namespace_prefix=\"loinc\"),\n",
    "               code=\"la30206-9\")   # Stand-alone evidence pathogenic\n",
    "    ],\n",
    "    name=\"Value set for 6.1.16 Clinical Annotation Level Of Evidence\",\n",
    "    description=(\"The level of evidence supporting the clinical annotation of \"\n",
    "                 \"the genetic variant.\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phenotypic Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6.2.1 Phenotypic Feature\n",
    "vs_6_2_1 = ValueSet(\n",
    "    elements=[HPO],  # HPO codes for phenotypic features\n",
    "    name=\"Value set for 6.2.1 Phenotypic Feature\",\n",
    "    description=(\"An observed physical and clinical characteristic encoded \"\n",
    "                 \"with HPO.\")\n",
    ")\n",
    "\n",
    "# 6.2.2 Determination Date\n",
    "vs_6_2_2 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 6.2.2 Determination Date\",\n",
    "    description=(\"The date on which the phenotypic feature was observed \"\n",
    "                 \"or recorded.\")\n",
    ")\n",
    "\n",
    "# 6.2.3 Status\n",
    "vs_6_2_3 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"410605003\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"723511001\")\n",
    "    ],\n",
    "    name=\"Value set for 6.2.3 Status\",\n",
    "    description=(\"The current status of the phenotypic feature, indicating \"\n",
    "                 \"whether it is confirmed or refuted.\")\n",
    ")\n",
    "\n",
    "# 6.2.4 Modifier\n",
    "vs_6_2_4 = ValueSet(\n",
    "    elements=[HPO, NCBITaxon, SNOMED],  # Grouped modifiers from HPO, NCBITaxon, SNOMED\n",
    "    name=\"Value set for 6.2.4 Modifier\",\n",
    "    description=(\"Further clinical modifiers to describe a specific \"\n",
    "                 \"phenotypic feature, such as severity or linked causative agents.\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Family History"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6.3.0 Pseudonym\n",
    "vs_6_3_0 = ValueSet(\n",
    "    elements=[str],  # A string for pseudonym\n",
    "    name=\"Value set for 6.3.0 Pseudonym\",\n",
    "    description=(\"If a pseudonym was already assigned to the specific family \"\n",
    "                 \"member, please enter it here to identify across records.\")\n",
    ")\n",
    "\n",
    "# 6.3.1 Propositus/-a\n",
    "vs_6_3_1 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373066001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373067005\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"1220561009\")\n",
    "    ],\n",
    "    name=\"Value set for 6.3.1 Propositus/-a\",\n",
    "    description=(\"Is the individual the first affected family member who seeks \"\n",
    "                 \"medical attention for a genetic disorder, leading to the \"\n",
    "                 \"diagnosis of other family members.\")\n",
    ")\n",
    "\n",
    "# 6.3.2 Relationship of the individual to the index case / propositus/a\n",
    "vs_6_3_2 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"65656005\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"9947008\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"83420006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"113160008\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"60614009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"73678001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"11286003\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"45929001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"2272004\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"62296006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"17945006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"1220561009\")\n",
    "    ],\n",
    "    name=\"Value set for 6.3.2 Relationship of the individual to the index case / propositus/a\",\n",
    "    description=(\"Specifies the familial relationship of the individual being \"\n",
    "                 \"evaluated to the index case or propositus/proposita.\")\n",
    ")\n",
    "\n",
    "# 6.3.3 Consanguinity\n",
    "vs_6_3_3 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373066001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373067005\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"1220561009\")\n",
    "    ],\n",
    "    name=\"Value set for 6.3.3 Consanguinity\",\n",
    "    description=(\"The presence of a biological relationship between parents who \"\n",
    "                 \"are related by blood, typically as first or second cousins.\")\n",
    ")\n",
    "\n",
    "# 6.3.4 Family Member Relationship\n",
    "vs_6_3_4 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"65656005\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"9947008\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"83420006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"113160008\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"60614009\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"73678001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"11286003\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"45929001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"2272004\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"62296006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"17945006\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"1220561009\")\n",
    "    ],\n",
    "    name=\"Value set for 6.3.4 Family Member Relationship\",\n",
    "    description=(\"Specifies the relationship of the selected family member to \"\n",
    "                 \"the patient.\")\n",
    ")\n",
    "\n",
    "# 6.3.5 Family Member Record Status\n",
    "vs_6_3_5 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"partial\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"completed\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"entered-in-error\"),\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"health-unknown\")\n",
    "    ],\n",
    "    name=\"Value set for 6.3.5 Family Member Record Status\",\n",
    "    description=(\"Specifies the record’s status of the family history of a \"\n",
    "                 \"specific family member.\")\n",
    ")\n",
    "\n",
    "# 6.3.6 Family Member Sex\n",
    "vs_6_3_6 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"248152002\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"248153007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"184115007\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"32570691000036108\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"1220561009\")\n",
    "    ],\n",
    "    name=\"Value set for 6.3.6 Family Member Sex\",\n",
    "    description=(\"Specifies the sex (or gender) of the specific family member. \"\n",
    "                 \"If possible, the sex assigned at birth should be selected.\")\n",
    ")\n",
    "\n",
    "# 6.3.7 Family Member Age\n",
    "vs_6_3_7 = ValueSet(\n",
    "    elements=[int],  # Integer for age\n",
    "    name=\"Value set for 6.3.7 Family Member Age\",\n",
    "    description=(\"Records the current age in full years of the selected family \"\n",
    "                 \"member.\")\n",
    ")\n",
    "\n",
    "# 6.3.8 Family Member Date of Birth\n",
    "vs_6_3_8 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 6.3.8 Family Member Date of Birth\",\n",
    "    description=(\"Records the date of birth of the selected family member.\")\n",
    ")\n",
    "\n",
    "# 6.3.9 Family Member Deceased\n",
    "vs_6_3_9 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373066001\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373067005\"),\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")\n",
    "    ],\n",
    "    name=\"Value set for 6.3.9 Family Member Deceased\",\n",
    "    description=(\"Indicates whether the selected family member is deceased.\")\n",
    ")\n",
    "\n",
    "# 6.3.10 Family Member Cause of Death [ICD10CM]\n",
    "vs_6_3_10 = ValueSet(\n",
    "    elements=[ICD10CM],  # ICD10CM for cause of death\n",
    "    name=\"Value set for 6.3.10 Family Member Cause of Death [ICD10CM]\",\n",
    "    description=(\"Records the cause of death of the selected deceased family member.\")\n",
    ")\n",
    "\n",
    "# 6.3.11 Family Member Deceased Age\n",
    "vs_6_3_11 = ValueSet(\n",
    "    elements=[int],  # Integer for age\n",
    "    name=\"Value set for 6.3.11 Family Member Deceased Age\",\n",
    "    description=(\"Records the age in full years of the selected family member \"\n",
    "                 \"at death.\")\n",
    ")\n",
    "\n",
    "# 6.3.12 Family Member Disease [MONDO]\n",
    "vs_6_3_12 = ValueSet(\n",
    "    elements=[MONDO],  # MONDO for family member disease\n",
    "    name=\"Value set for 6.3.12 Family Member Disease [MONDO]\",\n",
    "    description=(\"Indicates whether the selected family member is affected by the \"\n",
    "                 \"same RD as the individual or a different rare disease.\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Consent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7.1 Consent Status\n",
    "vs_7_1 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"draft\"),          # Pending\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"proposed\"),       # Proposed\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"active\"),         # Active\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"rejected\"),       # Rejected\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"inactive\"),       # Inactive\n",
    "        Coding(system=CodeSystem(name=\"HL7 FHIR\", namespace_prefix=\"hl7fhir\"),\n",
    "               code=\"entered-in-error\") # Entered in Error\n",
    "    ],\n",
    "    name=\"Value set for 7.1 Consent Status\",\n",
    "    description=(\"Indicates the current status of the consent.\")\n",
    ")\n",
    "\n",
    "# 7.2 Consent Date\n",
    "vs_7_2 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"7.2 Consent Date\",\n",
    "    description=(\"Records the date when the consent was given.\")\n",
    ")\n",
    "\n",
    "# 7.3 Health Policy Monitoring\n",
    "vs_7_3 = ValueSet(\n",
    "    elements=[str],  # Placeholder, as no specific coding is provided\n",
    "    name=\"7.3 Health Policy Monitoring\",\n",
    "    description=(\"The references to the policies that are included in this \"\n",
    "                 \"consent scope. Policies may be organisational, but are often \"\n",
    "                 \"defined jurisdictionally, or in law.\")\n",
    ")\n",
    "\n",
    "# 7.4 Agreement to be contacted for research purposes\n",
    "vs_7_4 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373066001\"),  # Yes\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373067005\"),  # No\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")   # Unknown\n",
    "    ],\n",
    "    name=\"Value set for 7.4 Agreement to be contacted for research purposes\",\n",
    "    description=(\"Indicates whether the patient agrees to be contacted for \"\n",
    "                 \"research purposes.\")\n",
    ")\n",
    "\n",
    "# 7.5 Consent to the reuse of data\n",
    "vs_7_5 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373066001\"),  # Yes\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373067005\"),  # No\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")   # Unknown\n",
    "    ],\n",
    "    name=\"7.5 Consent to the reuse of data\",\n",
    "    description=(\"Indicates whether the patient consents to the reuse of their \"\n",
    "                 \"data.\")\n",
    ")\n",
    "\n",
    "# 7.6 Biological sample\n",
    "vs_7_6 = ValueSet(\n",
    "    elements=[\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373066001\"),  # Yes\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"373067005\"),  # No\n",
    "        Coding(system=CodeSystem(name=\"SNOMED CT\", namespace_prefix=\"SNOMED\"),\n",
    "               code=\"261665006\")   # Unknown\n",
    "    ],\n",
    "    name=\"7.6 Biological sample\",\n",
    "    description=(\"Indicates whether a patient's biological sample is available \"\n",
    "                 \"for research.\")\n",
    ")\n",
    "\n",
    "# 7.7 Link to a biobank\n",
    "vs_7_7 = ValueSet(\n",
    "    elements=[str],  \n",
    "    name=\"7.7 Link to a biobank\",\n",
    "    description=(\"If there is a biological sample, this data element indicates \"\n",
    "                 \"the link to the biobank of the patient's biological sample.\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Disability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "vs_8_1 = ValueSet(\n",
    "    elements=[ICF],\n",
    "    name=\"Value set for 8.1 Classification of functioning / disability\",\n",
    "    description=(\"Classification of functioning and disability based on the \"\n",
    "                 \"International Classification of Functioning, Disability and \"\n",
    "                 \"Health (ICF).\"\n",
    "    )\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the Data Model object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards import DataModel, DataField"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "rarelink_cdm = DataModel(\n",
    "    data_model_name=\"RareLink CDM\",\n",
    "    resources=resources,  # Assuming resources are already defined\n",
    "    fields=[\n",
    "        # 1. Formal Criteria\n",
    "        DataField(section=\"1. Formal Criteria\", ordinal=\"1.1\", \n",
    "                   name=\"Pseudonym\", value_set=vs_1_1),\n",
    "        DataField(section=\"1. Formal Criteria\", ordinal=\"1.2\", \n",
    "                   name=\"Date of Admission\", value_set=vs_1_2),\n",
    "\n",
    "        # 2. Personal Information\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.1\", \n",
    "                   name=\"Date of Birth\", value_set=vs_2_1),\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.2\", \n",
    "                   name=\"Sex at Birth\", value_set=vs_2_2),\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.3\", \n",
    "                   name=\"Karyotypic Sex\", value_set=vs_2_3),\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.4\", \n",
    "                   name=\"Gender Identity\", value_set=vs_2_4),\n",
    "\n",
    "        # 3. Patient Status\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.1\", \n",
    "                   name=\"Vital Status\", value_set=vs_3_1),\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.2\", \n",
    "                   name=\"Time of Death\", value_set=vs_3_2),\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.3\", \n",
    "                   name=\"Cause of Death\", value_set=vs_3_3),\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.4\", \n",
    "                   name=\"Age Category\", value_set=vs_3_4),\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.5\", \n",
    "                   name=\"Length of Gestation at Birth\", value_set=vs_3_5),\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.6\", \n",
    "                   name=\"Undiagnosed RD Case\", value_set=vs_3_6),\n",
    "\n",
    "        # 4. Care Pathway\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.1\", \n",
    "                   name=\"Encounter Start\", value_set=vs_4_1),\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.2\", \n",
    "                   name=\"Encounter End\", value_set=vs_4_2),\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.3\", \n",
    "                   name=\"Encounter Status\", value_set=vs_4_3),\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.4\", \n",
    "                   name=\"Encounter Class\", value_set=vs_4_4),\n",
    "\n",
    "        # 5. Disease\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.1\", \n",
    "                   name=\"Disease\", value_set=vs_5_1),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.2\", \n",
    "                   name=\"Verification Status\", value_set=vs_5_2),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.3\", \n",
    "                   name=\"Age at Onset\", value_set=vs_5_3),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.4\", \n",
    "                   name=\"Date of Onset\", value_set=vs_5_4),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.5\", \n",
    "                   name=\"Age at Diagnosis\", value_set=vs_5_5),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.6\", \n",
    "                   name=\"Date of Diagnosis\", value_set=vs_5_6),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.7\", \n",
    "                   name=\"Body Site [SNOMED CT]\", value_set=vs_5_7),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.8\", \n",
    "                   name=\"Clinical Status\", value_set=vs_5_8),\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.9\", \n",
    "                   name=\"Severity\", value_set=vs_5_9),\n",
    "\n",
    "        # 6.1 Genetic Findings\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.1\", \n",
    "                   name=\"Genomic Diagnosis [MONDO]\", value_set=vs_6_1_1),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.2\", \n",
    "                   name=\"Progress Status of Interpretation\", value_set=vs_6_1_2),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.3\", \n",
    "                   name=\"Interpretation Status\", value_set=vs_6_1_3),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.4\", \n",
    "                   name=\"Structural Variant Analysis Method\", value_set=vs_6_1_4),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.5\", \n",
    "                   name=\"Reference Genome\", value_set=vs_6_1_5),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.6\", \n",
    "                   name=\"Genetic Mutation String\", value_set=vs_6_1_6),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.7\", \n",
    "                   name=\"Genomic DNA Change [g.HGVS]\", value_set=vs_6_1_7),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.8\", \n",
    "                   name=\"Sequence DNA Change [c.HGVS]\", value_set=vs_6_1_8),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.9\", \n",
    "                   name=\"Amino Acid Change [p.HGVS]\", value_set=vs_6_1_9),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.10\", \n",
    "                   name=\"Gene [HGNC-NR]\", value_set=vs_6_1_10),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.11\", \n",
    "                   name=\"Zygosity\", value_set=vs_6_1_11),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.12\", \n",
    "                   name=\"Genomic Source Class\", value_set=vs_6_1_12),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.13\", \n",
    "                   name=\"DNA Change Type\", value_set=vs_6_1_13),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.14\", \n",
    "                   name=\"Clinical Significance [ACMG]\", value_set=vs_6_1_14),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.15\", \n",
    "                   name=\"Therapeutic Actionability\", value_set=vs_6_1_15),\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.16\", \n",
    "                   name=\"Clinical Annotation Level Of Evidence\", value_set=vs_6_1_16),\n",
    "\n",
    "        # 6.2 Phenotypic Feature\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.1\", \n",
    "                   name=\"Phenotypic Feature\", value_set=vs_6_2_1),\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.2\", \n",
    "                   name=\"Determination Date\", value_set=vs_6_2_2),\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.3\", \n",
    "                   name=\"Status\", value_set=vs_6_2_3),\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.4\", \n",
    "                   name=\"Modifier\", value_set=vs_6_2_4),\n",
    "\n",
    "        # 6.3 Family History\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.0\", \n",
    "                   name=\"Family Member Pseudonym\", value_set=vs_6_3_0),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.1\", \n",
    "                   name=\"Propositus/-a\", value_set=vs_6_3_1),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.2\", \n",
    "                   name=\"Relationship of the individual to the index case / propositus/a\", \n",
    "                   value_set=vs_6_3_2),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.3\", \n",
    "                   name=\"Consanguinity\", value_set=vs_6_3_3),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.4\", \n",
    "                   name=\"Family Member Relationship\", value_set=vs_6_3_4),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.5\", \n",
    "                   name=\"Family Member Record Status\", value_set=vs_6_3_5),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.6\", \n",
    "                   name=\"Family Member Sex\", value_set=vs_6_3_6),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.7\", \n",
    "                   name=\"Family Member Age\", value_set=vs_6_3_7),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.8\", \n",
    "                   name=\"Family Member Date of Birth\", value_set=vs_6_3_8),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.9\", \n",
    "                   name=\"Family Member Deceased\", value_set=vs_6_3_9),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.10\", \n",
    "                   name=\"Family Member Cause of Death [ICD10CM]\", value_set=vs_6_3_10),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.11\", \n",
    "                   name=\"Family Member Deceased Age\", value_set=vs_6_3_11),\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.12\", \n",
    "                   name=\"Family Member Disease [MONDO]\", value_set=vs_6_3_12),\n",
    "\n",
    "        # 7. Consent\n",
    "        DataField(section=\"7. Consent\", ordinal=\"7.1\", \n",
    "                   name=\"Consent Status\", value_set=vs_7_1),\n",
    "        DataField(section=\"7. Consent\", ordinal=\"7.2\", \n",
    "                   name=\"Consent Date\", value_set=vs_7_2),\n",
    "        DataField(section=\"7. Consent\", ordinal=\"7.3\", \n",
    "                   name=\"Health Policy Monitoring\", value_set=vs_7_3),\n",
    "        DataField(section=\"7. Consent\", ordinal=\"7.4\", \n",
    "                   name=\"Agreement to be contacted for research purposes\", \n",
    "                   value_set=vs_7_4),\n",
    "        DataField(section=\"7. Consent\", ordinal=\"7.5\", \n",
    "                   name=\"Consent to the reuse of data\", \n",
    "                   value_set=vs_7_5),\n",
    "        DataField(section=\"7. Consent\", ordinal=\"7.6\", \n",
    "                   name=\"Biological sample\", \n",
    "                   value_set=vs_7_6),\n",
    "        DataField(section=\"7. Consent\", ordinal=\"7.7\", \n",
    "                   name=\"Link to a biobank\", \n",
    "                   value_set=vs_7_7),\n",
    "\n",
    "        # 8. Classification of functioning / disability\n",
    "        DataField(section=\"8. Classification\", ordinal=\"8.1\", \n",
    "                   name=\"Classification of functioning / disability\", \n",
    "                   value_set=vs_8_1),\n",
    "    ]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
