{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing the RareLink RD CDM for the Phenopacket Mapper\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RD CDM v2.0: 'res/RD CDM v2.0.xlsx'\n",
    "\n",
    "\n",
    "RareLink Data Dictionary: '../res/RareLink_v2.0_DataDictionary.csv'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 - Defining the RD CDM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 Resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards.code_system import ORDO, HPO, MONDO, OMIM, LOINC, HGNC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards.code_system import CodeSystem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NCIT = CodeSystem(\n",
    "    name='National Cancer Institute Thesaurus (NCIT)', \n",
    "    namespace_prefix='ncit', \n",
    "    url='https://ncithesaurus.org/'\n",
    ")\n",
    "\n",
    "UO = CodeSystem(\n",
    "    name='Units of Measurement Ontology (UO)', \n",
    "    namespace_prefix='uo', \n",
    "    url='http://purl.obolibrary.org/obo/uo.owl'\n",
    ")\n",
    "\n",
    "NCBITaxon = CodeSystem(\n",
    "    name='NCBI Taxonomy', \n",
    "    namespace_prefix='ncbitaxon', \n",
    "    url='https://www.ncbi.nlm.nih.gov/taxonomy'\n",
    ")\n",
    "\n",
    "GENO = CodeSystem(\n",
    "    name='GENO: The Genotype Ontology', \n",
    "    namespace_prefix='geno', \n",
    "    url='http://www.genoontology.org/'\n",
    ")\n",
    "\n",
    "SO = CodeSystem(\n",
    "    name='Sequence Ontology (SO)', \n",
    "    namespace_prefix='so', \n",
    "    url='http://www.sequenceontology.org/'\n",
    ")\n",
    "\n",
    "ICD10CM = CodeSystem(\n",
    "    name='ICD-10-CM', \n",
    "    namespace_prefix='icd10cm', \n",
    "    url='https://www.cdc.gov/nchs/icd/icd10cm.htm'\n",
    ")\n",
    "\n",
    "SNOMED = CodeSystem(\n",
    "    name='SNOMED CT', \n",
    "    namespace_prefix='snomed', \n",
    "    url='https://www.snomed.org/snomed-ct'\n",
    ")\n",
    "\n",
    "ICD11 = CodeSystem(\n",
    "    name='ICD-11', \n",
    "    namespace_prefix='icd11', \n",
    "    url='https://icd.who.int/en'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "resources = [\n",
    "    HPO.set_version(\"2024-08-13\"),\n",
    "    MONDO.set_version(\"2024-09-03\"),\n",
    "    OMIM.set_version(\"2024-09-12\"),\n",
    "    ORDO.set_version(\"2024-09-12\"),\n",
    "    NCBITaxon.set_version(\"2024-07-03\"),\n",
    "    LOINC.set_version(\"2.78\"),\n",
    "    HGNC.set_version(\"2024-08-23\"),\n",
    "    GENO.set_version(\"2023-10-08\"),\n",
    "    NCIT.set_version(\"24.04e\"),\n",
    "    SO.set_version(\"2.6\"),\n",
    "    ICD10CM.set_version(\"2024-09-01\"),\n",
    "    SNOMED.set_version(\"2024-09-01\"),\n",
    "    ICD11.set_version(\"2024-09-01\"),\n",
    "]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.2 Fields of the data model and their value sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards.value_set import ValueSet\n",
    "from phenopacket_mapper.data_standards import Date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Formal Criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vs_1_1 = ValueSet(\n",
    "    elements=[str],\n",
    "    name=\"Value set for 1.1 Pseudonym\",\n",
    "    description=\"Value set for field 1.1 Pseudonym of the RareLink data model in section rarelink_1_formal_criteria. This field represents the (local) patient-related Identification code.\"\n",
    ")\n",
    "\n",
    "vs_1_2 = ValueSet(\n",
    "    elements=[Date],\n",
    "    name=\"Value set for 1.2 Date of Admission\",\n",
    "    description=\"Value set for field 1.2 Date of Admission of the RareLink data model in section rarelink_1_formal_criteria. This field represents the date of admission or data capture of the individual.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Personal Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vs_2_1 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 2.1 Date of Birth\",\n",
    "    description=\"Value set for field 2.1 Date of Birth of the RareLink data model in section rarelink_2_personal_information. The individual's date of birth. If the exact month or day is allowed to be captured or not known, select the 1st day of the month or the 1st month of the year, respectively.\"\n",
    ")\n",
    "\n",
    "vs_2_2 = ValueSet(\n",
    "    elements=[\"snomed_248152002\", \"snomed_248153007\", \"snomed_184115007\", \"snomed_32570691000036108\", \"snomed_1220561009\"],\n",
    "    name=\"Value set for 2.2 Sex at Birth\",\n",
    "    description=\"Value set for field 2.2 Sex at Birth of the RareLink data model in section rarelink_2_personal_information. The individual's sex that was assigned at birth.\"\n",
    ")\n",
    "\n",
    "vs_2_3 = ValueSet(\n",
    "    elements=[\"snomed_261665006\", \"snomed_734875008\", \"snomed_734876009\", \"snomed_80427008\", \"snomed_65162001\", \"snomed_35111009\", \"snomed_403760006\", \"snomed_78317008\", \"snomed_10567003\", \"snomed_48930007\", \"snomed_74964007\"],\n",
    "    name=\"Value set for 2.3 Karyotypic Sex\",\n",
    "    description=\"Value set for field 2.3 Karyotypic Sex of the RareLink data model in section rarelink_2_personal_information. The chromosomal sex of an individual.\"\n",
    ")\n",
    "\n",
    "vs_2_4 = ValueSet(\n",
    "    elements=[\"snomed_446141000124107\", \"snomed_446151000124109\", \"snomed_394743007\", \"snomed_33791000087105\", \"snomed_1220561009\"],\n",
    "    name=\"Value set for 2.4 Gender Identity\",\n",
    "    description=\"Value set for field 2.4 Gender Identity of the RareLink data model in section rarelink_2_personal_information. The self-assigned gender of the individual.\"\n",
    ")\n",
    "\n",
    "vs_2_5 = ValueSet(\n",
    "    elements=[\"ISO 3166 2-letter code\", \"ISO 3166 3-letter code\"],  # Based on ISO 3166 country codes\n",
    "    name=\"Value set for 2.5 Country of Birth\",\n",
    "    description=\"Value set for field 2.5 Country of Birth of the RareLink data model in section rarelink_2_personal_information. The individual's country of birth. Please enter a two or three letter ISO3166 code.\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Formal Criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.1 Vital Status\n",
    "vs_3_1 = ValueSet(\n",
    "    elements=[\"snomed_438949009\", \"snomed_419099009\", \"snomed_399307001\", \"snomed_185924006\", \"snomed_261665006\"],\n",
    "    name=\"Value set for 3.1 Vital Status\",\n",
    "    description=\"Value set for field 3.1 Vital Status of the RareLink data model in section rarelink_3_patient_status. The individual’s general clinical status or vital status.\"\n",
    ")\n",
    "\n",
    "# 3.2 Time of Death\n",
    "vs_3_2 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 3.2 Time of Death\",\n",
    "    description=\"Value set for field 3.2 Time of Death of the RareLink data model in section rarelink_3_patient_status. If deceased, the individual’s date of death. If the specific month or day is not known, select the 1st day of the month or the 1st month of the year, respectively.\"\n",
    ")\n",
    "\n",
    "# 3.3 Cause of Death [ICD10CM]\n",
    "vs_3_3 = ValueSet(\n",
    "    elements=[ICD10CM],  # Based on ICD10CM coding\n",
    "    name=\"Value set for 3.3 Cause of Death [ICD10CM]\",\n",
    "    description=\"Value set for field 3.3 Cause of Death [ICD10CM] of the RareLink data model in section rarelink_3_patient_status. If deceased, the individual’s primary cause of death (i.e. according to the death certificate).\"\n",
    ")\n",
    "\n",
    "# 3.4 Age Category\n",
    "vs_3_4 = ValueSet(\n",
    "    elements=[\"snomed_3658006\", \"snomed_713153009\", \"snomed_255398004\", \"snomed_263659003\", \"snomed_41847000\", \"snomed_303112003\", \"snomed_419099009\", \"snomed_261665006\"],\n",
    "    name=\"Value set for 3.4 Age Category\",\n",
    "    description=\"Value set for field 3.4 Age Category of the RareLink data model in section rarelink_3_patient_status. The individual's age category at the time of data capture (1.2).\"\n",
    ")\n",
    "\n",
    "# 3.5 Length of Gestation at Birth [weeks+days]\n",
    "vs_3_5 = ValueSet(\n",
    "    elements=[str],  # Custom format for weeks and days\n",
    "    name=\"Value set for 3.5 Length of Gestation at Birth [weeks+days]\",\n",
    "    description=\"Value set for field 3.5 Length of Gestation at Birth of the RareLink data model in section rarelink_3_patient_status. The duration of the pregnancy in weeks and days, from the first day of the last menstrual period to the day of delivery, formatted as XX+X (weeks+days).\"\n",
    ")\n",
    "\n",
    "# 3.6 Undiagnosed RD Case\n",
    "vs_3_6 = ValueSet(\n",
    "    elements=[\"snomed_373066001\", \"snomed_373067005\"],\n",
    "    name=\"Value set for 3.6 Undiagnosed RD Case\",\n",
    "    description=\"Value set for field 3.6 Undiagnosed RD Case of the RareLink data model in section rarelink_3_patient_status. Identifies cases where an RD diagnosis has not been established.\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Care Pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4.1 Encounter Start\n",
    "vs_4_1 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 4.1 Encounter Start\",\n",
    "    description=\"Value set for field 4.1 Encounter Start of the RareLink data model in section rarelink_4_care_pathway. The beginning of an encounter of the individual. If the specific month or day is not known, select the 1st day of the month or the 1st month of the year, respectively.\"\n",
    ")\n",
    "\n",
    "# 4.2 Encounter End\n",
    "vs_4_2 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 4.2 Encounter End\",\n",
    "    description=\"Value set for field 4.2 Encounter End of the RareLink data model in section rarelink_4_care_pathway. The end of an encounter of the individual. If the specific month or day is not known, select the 1st day of the month or the 1st month of the year, respectively.\"\n",
    ")\n",
    "\n",
    "# 4.3 Encounter Status\n",
    "vs_4_3 = ValueSet(\n",
    "    elements=[\"hl7fhir_planned\", \"hl7fhir_arrived\", \"hl7fhir_triaged\", \"hl7fhir_in-progress\", \"hl7fhir_onleave\", \"hl7fhir_finished\", \"hl7fhir_cancelled\", \"hl7fhir_entered-in-error\", \"hl7fhir_unknown\"],\n",
    "    name=\"Value set for 4.3 Encounter Status\",\n",
    "    description=\"Value set for field 4.3 Encounter Status of the RareLink data model in section rarelink_4_care_pathway. The status of an encounter of the individual at the time of data capture.\"\n",
    ")\n",
    "\n",
    "# 4.4 Encounter Class\n",
    "vs_4_4 = ValueSet(\n",
    "    elements=[\"hl7fhir_amb\", \"hl7fhir_imp\", \"hl7fhir_obsenc\", \"hl7fhir_emer\", \"hl7fhir_vr\", \"hl7fhir_hh\", \"hl7fhir_rdc\", \"snomed_261665006\"],\n",
    "    name=\"Value set for 4.4 Encounter Class\",\n",
    "    description=\"Value set for field 4.4 Encounter Class of the RareLink data model in section rarelink_4_care_pathway. The class of an encounter of the individual at the time of data capture.\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Disease\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5.1 Disease [MONDO]\n",
    "vs_5_1 = ValueSet(\n",
    "    elements=[MONDO, ORDO, ICD10CM, ICD11, OMIM], \n",
    "    name=\"Value set for 5.1 Disease [MONDO, ORDO, ICD10CM, ICD11, OMIM]\",\n",
    "    description=\"Value set for field 5.1 Disease [MONDO, ORDO, ICD10CM, ICD11, OMIM] of the RareLink data model in section rarelink_5_disease. Please look up the disease code for the disease that the individual was affected by.\"\n",
    ")\n",
    "\n",
    "# 5.2 Verification Status\n",
    "vs_5_2 = ValueSet(\n",
    "    elements=[\"hl7fhir_unconfirmed\", \"hl7fhir_provisional\", \"hl7fhir_differential\", \"hl7fhir_confirmed\", \"hl7fhir_refuted\", \"hl7fhir_entered-in-error\"],\n",
    "    name=\"Value set for 5.2 Verification Status\",\n",
    "    description=\"Value set for field 5.2 Verification Status of the RareLink data model in section rarelink_5_disease. The verification status of the disease.\"\n",
    ")\n",
    "\n",
    "# 5.3 Age at Onset\n",
    "vs_5_3 = ValueSet(\n",
    "    elements=[\"snomed_118189007\", \"snomed_3950001\", \"snomed_410672004\", \"snomed_261665006\"],\n",
    "    name=\"Value set for 5.3 Age at Onset\",\n",
    "    description=\"Value set for field 5.3 Age at Onset of the RareLink data model in section rarelink_5_disease. The age at the onset of the first symptoms or signs of the disease.\"\n",
    ")\n",
    "\n",
    "# 5.4 Date of Onset\n",
    "vs_5_4 = ValueSet(\n",
    "    elements=[Date],  \n",
    "    name=\"Value set for 5.4 Date of Onset\",\n",
    "    description=\"Value set for field 5.4 Date of Onset of the RareLink data model in section rarelink_5_disease. The date at onset of first symptoms or signs of the disease. If the specific month or day is not known, select the 1st day of the month or the 1st month of the year, respectively.\"\n",
    ")\n",
    "\n",
    "# 5.5 Age at Diagnosis\n",
    "vs_5_5 = ValueSet(\n",
    "    elements=[\"snomed_118189007\", \"snomed_3950001\", \"snomed_410672004\", \"snomed_261665006\"],\n",
    "    name=\"Value set for 5.5 Age at Diagnosis\",\n",
    "    description=\"Value set for field 5.5 Age at Diagnosis of the RareLink data model in section rarelink_5_disease. The individual’s age when the diagnosis was made.\"\n",
    ")\n",
    "\n",
    "# 5.6 Date of Diagnosis\n",
    "vs_5_6 = ValueSet(\n",
    "    elements=[Date], \n",
    "    name=\"Value set for 5.6 Date of Diagnosis\",\n",
    "    description=\"Value set for field 5.6 Date of Diagnosis of the RareLink data model in section rarelink_5_disease. The date on which the disease was determined. If the specific month or day is not known, select the 1st day of the month or the 1st month of the year, respectively.\"\n",
    ")\n",
    "\n",
    "# 5.7 Body Site [SNOMED CT]\n",
    "vs_5_7 = ValueSet(\n",
    "    elements=[SNOMED], \n",
    "    name=\"Value set for 5.7 Body Site [SNOMED CT]\",\n",
    "    description=\"Value set for field 5.7 Body Site [SNOMED CT] of the RareLink data model in section rarelink_5_disease. The specific body site affected by disease is encoded using all descendants of SCT Body Structure (123037004).\"\n",
    ")\n",
    "\n",
    "# 5.8 Clinical Status\n",
    "vs_5_8 = ValueSet(\n",
    "    elements=[\"hl7fhir_active\", \"hl7fhir_recurrence\", \"hl7fhir_relapse\", \"hl7fhir_inactive\", \"hl7fhir_remission\", \"hl7fhir_resolved\"],\n",
    "    name=\"Value set for 5.8 Clinical Status\",\n",
    "    description=\"Value set for field 5.8 Clinical Status of the RareLink data model in section rarelink_5_disease. The clinical status of the disease indicates whether it is active, inactive, or resolved.\"\n",
    ")\n",
    "\n",
    "# 5.9 Severity\n",
    "vs_5_9 = ValueSet(\n",
    "    elements=[\"snomed_24484000\", \"snomed_6736007\", \"snomed_255604002\"],\n",
    "    name=\"Value set for 5.9 Severity\",\n",
    "    description=\"Value set for field 5.9 Severity of the RareLink data model in section rarelink_5_disease. The severity of the disease is categorized by clinical evaluation.\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Genetic Finddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6.1.1 Genomic Diagnosis [MONDO, OMIM]\n",
    "vs_6_1_1 = ValueSet(\n",
    "    elements=[MONDO, OMIM],\n",
    "    name=\"Value set for 6.1.1 Genomic Diagnosis [MONDO]\",\n",
    "    description=\"Value set for field 6.1.1 Genomic Diagnosis [MONDO] of the RareLink data model in section rarelink_6_1_genetic_findings. The genetic finding of a variant can be linked to a disease in (5.1) if the same MONDO code is used.\"\n",
    ")\n",
    "\n",
    "# 6.1.2 Progress Status of Interpretation\n",
    "vs_6_1_2 = ValueSet(\n",
    "    elements=[\"ga4gh_unknown_progress\", \"ga4gh_in_progress\", \"ga4gh_completed\", \"ga4gh_solved\", \"ga4gh_unsolved\"],\n",
    "    name=\"Value set for 6.1.2 Progress Status of Interpretation\",\n",
    "    description=\"Value set for field 6.1.2 Progress Status of Interpretation of the RareLink data model in section rarelink_6_1_genetic_findings. The interpretation has a ProgressStatus that refers to the status of the attempted diagnosis.\"\n",
    ")\n",
    "\n",
    "# 6.1.3 Interpretation Status\n",
    "vs_6_1_3 = ValueSet(\n",
    "    elements=[\"ga4gh_unknown_status\", \"ga4gh_rejected\", \"ga4gh_candidate\", \"ga4gh_contributory\", \"ga4gh_causative\"],\n",
    "    name=\"Value set for 6.1.3 Interpretation Status\",\n",
    "    description=\"Value set for field 6.1.3 Interpretation Status of the RareLink data model in section rarelink_6_1_genetic_findings. An enumeration that describes the conclusion made about the genomic interpretation.\"\n",
    ")\n",
    "\n",
    "# 6.1.4 Structural Variant Analysis Method\n",
    "vs_6_1_4 = ValueSet(\n",
    "    elements=[\"loinc_la26406-1\", \"loinc_la26404-6\", \"loinc_la26418-6\", \"loinc_la26419-4\", \"loinc_la26400-4\", \"loinc_la26813-8\", \"loinc_la26810-4\", \"loinc_la26398-0\", \"loinc_la26415-2\", \"loinc_la46-8\"],\n",
    "    name=\"Value set for 6.1.4 Structural Variant Analysis Method\",\n",
    "    description=\"Value set for field 6.1.4 Structural Variant Analysis Method of the RareLink data model in section rarelink_6_1_genetic_findings. The method used to analyze structural variants in the genome.\"\n",
    ")\n",
    "\n",
    "# 6.1.5 Reference Genome\n",
    "vs_6_1_5 = ValueSet(\n",
    "    elements=[\"loinc_la14032-9\", \"loinc_la14029-5\", \"loinc_la14030-3\", \"loinc_la14031-1\", \"loinc_la26806-2\"],\n",
    "    name=\"Value set for 6.1.5 Reference Genome\",\n",
    "    description=\"Value set for field 6.1.5 Reference Genome of the RareLink data model in section rarelink_6_1_genetic_findings. The reference genome used for analyzing the genetic variant.\"\n",
    ")\n",
    "\n",
    "# 6.1.6 Genetic Mutation String\n",
    "vs_6_1_6 = ValueSet(\n",
    "    elements=[str],  # HGVS strings for genetic mutations\n",
    "    name=\"Value set for 6.1.6 Genetic Mutation String\",\n",
    "    description=\"Value set for field 6.1.6 Genetic Mutation String of the RareLink data model in section rarelink_6_1_genetic_findings. An unvalidated (HGVS) string that describes the variant change.\"\n",
    ")\n",
    "\n",
    "# 6.1.7 Genomic DNA Change [g.HGVS]\n",
    "vs_6_1_7 = ValueSet(\n",
    "    elements=[str],  # HGVS strings for genomic DNA change\n",
    "    name=\"Value set for 6.1.7 Genomic DNA Change [g.HGVS]\",\n",
    "    description=\"Value set for field 6.1.7 Genomic DNA Change [g.HGVS] of the RareLink data model in section rarelink_6_1_genetic_findings. The specific change in the genomic DNA sequence encoded with a validated g.HGVS expression.\"\n",
    ")\n",
    "\n",
    "# 6.1.8 Sequence DNA Change [c.HGVS]\n",
    "vs_6_1_8 = ValueSet(\n",
    "    elements=[str],  # HGVS strings for DNA sequence changes\n",
    "    name=\"Value set for 6.1.8 Sequence DNA Change [c.HGVS]\",\n",
    "    description=\"Value set for field 6.1.8 Sequence DNA Change [c.HGVS] of the RareLink data model in section rarelink_6_1_genetic_findings. The specific change in the DNA sequence at the nucleotide level with a validated c.HGVS expression.\"\n",
    ")\n",
    "\n",
    "# 6.1.9 Amino Acid Change [p.HGVS]\n",
    "vs_6_1_9 = ValueSet(\n",
    "    elements=[str],  # HGVS strings for amino acid changes\n",
    "    name=\"Value set for 6.1.9 Amino Acid Change [p.HGVS]\",\n",
    "    description=\"Value set for field 6.1.9 Amino Acid Change [p.HGVS] of the RareLink data model in section rarelink_6_1_genetic_findings. The specific change in the amino acid sequence resulting from a genetic variant as a validated p.HGVS expression.\"\n",
    ")\n",
    "\n",
    "# 6.1.10 Gene [HGNC-NR]\n",
    "vs_6_1_10 = ValueSet(\n",
    "    elements=[HGNC],  # Gene codes\n",
    "    name=\"Value set for 6.1.10 Gene [HGNC-NR]\",\n",
    "    description=\"Value set for field 6.1.10 Gene [HGNC-NR] of the RareLink data model in section rarelink_6_1_genetic_findings. The specific gene or genes that were analyzed or identified in the study.\"\n",
    ")\n",
    "\n",
    "# 6.1.11 Zygosity\n",
    "vs_6_1_11 = ValueSet(\n",
    "    elements=[\"loinc_la6705-3\", \"loinc_la6706-1\", \"loinc_la26217-2\", \"loinc_la26220-6\", \"loinc_la6707-9\", \"loinc_la6703-8\", \"loinc_la6704-6\", \"loinc_53034_5_other\"],\n",
    "    name=\"Value set for 6.1.11 Zygosity\",\n",
    "    description=\"Value set for field 6.1.11 Zygosity of the RareLink data model in section rarelink_6_1_genetic_findings. The zygosity of the genetic variant.\"\n",
    ")\n",
    "\n",
    "# 6.1.12 Genomic Source Class\n",
    "vs_6_1_12 = ValueSet(\n",
    "    elements=[\"loinc_la6683-2\", \"loinc_la6684-0\", \"loinc_la10429-1\", \"loinc_la18194-3\", \"loinc_la18195-0\", \"loinc_la18196-8\", \"loinc_la18197-6\", \"loinc_la26807-0\"],\n",
    "    name=\"Value set for 6.1.12 Genomic Source Class\",\n",
    "    description=\"Value set for field 6.1.12 Genomic Source Class of the RareLink data model in section rarelink_6_1_genetic_findings. The classification of the genomic source, such as germline, somatic, or other origins.\"\n",
    ")\n",
    "\n",
    "# 6.1.13 DNA Change Type\n",
    "vs_6_1_13 = ValueSet(\n",
    "    elements=[\"loinc_la9658-1\", \"loinc_la6692-3\", \"loinc_la6686-5\", \"loinc_la6687-3\", \"loinc_la6688-1\", \"loinc_la6689-9\", \"loinc_la6690-7\", \"loinc_48019_4_other\"],\n",
    "    name=\"Value set for 6.1.13 DNA Change Type\",\n",
    "    description=\"Value set for field 6.1.13 DNA Change Type of the RareLink data model in section rarelink_6_1_genetic_findings. The variant’s type of DNA change, such as point mutation, deletion, insertion, or other types.\"\n",
    ")\n",
    "\n",
    "# 6.1.14 Clinical Significance [ACMG]\n",
    "vs_6_1_14 = ValueSet(\n",
    "    elements=[\"loinc_la6668-3\", \"loinc_la26332-9\", \"loinc_la26333-7\", \"loinc_la26334-5\", \"loinc_la6675-8\", \"loinc_la4489-6\"],\n",
    "    name=\"Value set for 6.1.14 Clinical Significance [ACMG]\",\n",
    "    description=\"Value set for field 6.1.14 Clinical Significance [ACMG] of the RareLink data model in section rarelink_6_1_genetic_findings. The clinical significance of the genetic variant, indicating its impact on health and disease.\"\n",
    ")\n",
    "\n",
    "\n",
    "# 6.1.15 Therapeutic Actionability\n",
    "vs_6_1_15 = ValueSet(\n",
    "    elements=[\"ga4gh_unknown_actionability\", \"ga4gh_not_actionable\", \"ga4gh_actionable\"],\n",
    "    name=\"Value set for 6.1.15 Therapeutic Actionability\",\n",
    "    description=\"Value set for field 6.1.15 Therapeutic Actionability of the RareLink data model in section rarelink_6_1_genetic_findings. This field flags the variant as being a candidate for treatment/clinical intervention, which could improve the clinical outcome.\"\n",
    ")\n",
    "\n",
    "# 6.1.16 Clinical Annotation Level Of Evidence\n",
    "vs_6_1_16 = ValueSet(\n",
    "    elements=[\"loinc_la30200-2\", \"loinc_la30201-0\", \"loinc_la30202-8\", \"loinc_la30203-6\", \"loinc_la30204-4\", \"loinc_la30205-1\", \"loinc_la30206-9\"],\n",
    "    name=\"Value set for 6.1.16 Clinical Annotation Level Of Evidence\",\n",
    "    description=\"Value set for field 6.1.16 Clinical Annotation Level Of Evidence of the RareLink data model in section rarelink_6_1_genetic_findings. The level of evidence supporting the clinical annotation of the genetic variant.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phenotypic Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6.2.1 Phenotypic Feature\n",
    "vs_6_2_1 = ValueSet(\n",
    "    elements=[HPO],  # HPO codes for phenotypic features\n",
    "    name=\"Value set for 6.2.1 Phenotypic Feature\",\n",
    "    description=\"Value set for field 6.2.1 Phenotypic Feature of the RareLink data model in section rarelink_6_2_phenotypic_feature. An observed physical and clinical characteristic encoded with HPO.\"\n",
    ")\n",
    "\n",
    "# 6.2.2 Determination Date\n",
    "vs_6_2_2 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 6.2.2 Determination Date\",\n",
    "    description=\"Value set for field 6.2.2 Determination Date of the RareLink data model in section rarelink_6_2_phenotypic_feature. The date on which the phenotypic feature was observed or recorded.\"\n",
    ")\n",
    "\n",
    "# 6.2.3 Status\n",
    "vs_6_2_3 = ValueSet(\n",
    "    elements=[\"snomed_410605003\", \"snomed_723511001\"],\n",
    "    name=\"Value set for 6.2.3 Status\",\n",
    "    description=\"Value set for field 6.2.3 Status of the RareLink data model in section rarelink_6_2_phenotypic_feature. The current status of the phenotypic feature, indicating whether it is confirmed or refuted.\"\n",
    ")\n",
    "\n",
    "# 6.2.4 Modifier\n",
    "vs_6_2_4 = ValueSet(\n",
    "    elements=[HPO, NCBITaxon, SNOMED],  # Grouped modifiers from HPO, NCBITaxon, SNOMED\n",
    "    name=\"Value set for 6.2.4 Modifier\",\n",
    "    description=\"Value set for field 6.2.4 Modifier of the RareLink data model in section rarelink_6_2_phenotypic_feature. Further clinical modifiers to describe a specific phenotypic feature, such as severity or linked causative agents.\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Family History"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6.3.0 Pseudonym\n",
    "vs_6_3_0 = ValueSet(\n",
    "    elements=[str],  # A string for pseudonym\n",
    "    name=\"Value set for 6.3.0 Pseudonym\",\n",
    "    description=\"Value set for field 6.3.0 Pseudonym of the RareLink data model in section rarelink_6_3_family_history. If a pseudonym was already assigned to the specific family member, please enter it here to identify across records.\"\n",
    ")\n",
    "\n",
    "# 6.3.1 Propositus/-a\n",
    "vs_6_3_1 = ValueSet(\n",
    "    elements=[\"snomed_373066001\", \"snomed_373067005\", \"snomed_261665006\", \"snomed_1220561009\"],\n",
    "    name=\"Value set for 6.3.1 Propositus/-a\",\n",
    "    description=\"Value set for field 6.3.1 Propositus/-a of the RareLink data model in section rarelink_6_3_family_history. Is the individual the first affected family member who seeks medical attention for a genetic disorder, leading to the diagnosis of other family members.\"\n",
    ")\n",
    "\n",
    "# 6.3.2 Relationship of the individual to the index case / propositus/a\n",
    "vs_6_3_2 = ValueSet(\n",
    "    elements=[\"snomed_65656005\", \"snomed_9947008\", \"snomed_83420006\", \"snomed_113160008\", \"snomed_60614009\", \"snomed_73678001\", \"snomed_11286003\", \"snomed_45929001\", \"snomed_2272004\", \"snomed_62296006\", \"snomed_17945006\", \"snomed_1220561009\"],\n",
    "    name=\"Value set for 6.3.2 Relationship of the individual to the index case / propositus/a\",\n",
    "    description=\"Value set for field 6.3.2 Relationship of the individual to the index case / propositus/a of the RareLink data model in section rarelink_6_3_family_history. Specifies the familial relationship of the individual being evaluated to the index case or propositus/proposita.\"\n",
    ")\n",
    "\n",
    "# 6.3.3 Consanguinity\n",
    "vs_6_3_3 = ValueSet(\n",
    "    elements=[\"snomed_373066001\", \"snomed_373067005\", \"snomed_261665006\", \"snomed_1220561009\"],\n",
    "    name=\"Value set for 6.3.3 Consanguinity\",\n",
    "    description=\"Value set for field 6.3.3 Consanguinity of the RareLink data model in section rarelink_6_3_family_history. The presence of a biological relationship between parents who are related by blood, typically as first or second cousins.\"\n",
    ")\n",
    "\n",
    "# 6.3.4 Family Member Relationship\n",
    "vs_6_3_4 = ValueSet(\n",
    "    elements=[\"snomed_65656005\", \"snomed_9947008\", \"snomed_83420006\", \"snomed_113160008\", \"snomed_60614009\", \"snomed_73678001\", \"snomed_11286003\", \"snomed_45929001\", \"snomed_2272004\", \"snomed_62296006\", \"snomed_17945006\", \"snomed_1220561009\"],\n",
    "    name=\"Value set for 6.3.4 Family Member Relationship\",\n",
    "    description=\"Value set for field 6.3.4 Family Member Relationship of the RareLink data model in section rarelink_6_3_family_history. Specifies the relationship of the selected family member to the patient.\"\n",
    ")\n",
    "\n",
    "# 6.3.5 Family Member Record Status\n",
    "vs_6_3_5 = ValueSet(\n",
    "    elements=[\"hl7fhir_partial\", \"hl7fhir_completed\", \"hl7fhir_entered-in-error\", \"hl7fhir_health-unknown\"],\n",
    "    name=\"Value set for 6.3.5 Family Member Record Status\",\n",
    "    description=\"Value set for field 6.3.5 Family Member Record Status of the RareLink data model in section rarelink_6_3_family_history. Specifies the record’s status of the family history of a specific family member.\"\n",
    ")\n",
    "\n",
    "# 6.3.6 Family Member Sex\n",
    "vs_6_3_6 = ValueSet(\n",
    "    elements=[\"snomed_248152002\", \"snomed_248153007\", \"snomed_184115007\", \"snomed_32570691000036108\", \"snomed_1220561009\"],\n",
    "    name=\"Value set for 6.3.6 Family Member Sex\",\n",
    "    description=\"Value set for field 6.3.6 Family Member Sex of the RareLink data model in section rarelink_6_3_family_history. Specifies the sex (or gender) of the specific family member. If possible, the sex assigned at birth should be selected.\"\n",
    ")\n",
    "\n",
    "# 6.3.7 Family Member Age\n",
    "vs_6_3_7 = ValueSet(\n",
    "    elements=[int],  # Integer for age\n",
    "    name=\"Value set for 6.3.7 Family Member Age\",\n",
    "    description=\"Value set for field 6.3.7 Family Member Age of the RareLink data model in section rarelink_6_3_family_history. Records the current age in full years of the selected family member.\"\n",
    ")\n",
    "\n",
    "# 6.3.8 Family Member Date of Birth\n",
    "vs_6_3_8 = ValueSet(\n",
    "    elements=[Date],  # Based on the \"date_ymd\" format\n",
    "    name=\"Value set for 6.3.8 Family Member Date of Birth\",\n",
    "    description=\"Value set for field 6.3.8 Family Member Date of Birth of the RareLink data model in section rarelink_6_3_family_history. Records the date of birth of the selected family member.\"\n",
    ")\n",
    "\n",
    "# 6.3.9 Family Member Deceased\n",
    "vs_6_3_9 = ValueSet(\n",
    "    elements=[\"snomed_373066001\", \"snomed_373067005\", \"snomed_261665006\"],\n",
    "    name=\"Value set for 6.3.9 Family Member Deceased\",\n",
    "    description=\"Value set for field 6.3.9 Family Member Deceased of the RareLink data model in section rarelink_6_3_family_history. Indicates whether the selected family member is deceased.\"\n",
    ")\n",
    "\n",
    "# 6.3.10 Family Member Cause of Death [ICD10CM]\n",
    "vs_6_3_10 = ValueSet(\n",
    "    elements=[ICD10CM],  # ICD10CM for cause of death\n",
    "    name=\"Value set for 6.3.10 Family Member Cause of Death [ICD10CM]\",\n",
    "    description=\"Value set for field 6.3.10 Family Member Cause of Death [ICD10CM] of the RareLink data model in section rarelink_6_3_family_history. Records the cause of death of the selected deceased family member.\"\n",
    ")\n",
    "\n",
    "# 6.3.11 Family Member Deceased Age\n",
    "vs_6_3_11 = ValueSet(\n",
    "    elements=[int],  # Integer for age\n",
    "    name=\"Value set for 6.3.11 Family Member Deceased Age\",\n",
    "    description=\"Value set for field 6.3.11 Family Member Deceased Age of the RareLink data model in section rarelink_6_3_family_history. Records the age in full years of the selected family member at death.\"\n",
    ")\n",
    "\n",
    "# 6.3.12 Family Member Disease [MONDO]\n",
    "vs_6_3_12 = ValueSet(\n",
    "    elements=[MONDO],  # MONDO for family member disease\n",
    "    name=\"Value set for 6.3.12 Family Member Disease [MONDO]\",\n",
    "    description=\"Value set for field 6.3.12 Family Member Disease [MONDO] of the RareLink data model in section rarelink_6_3_family_history. Indicates whether the selected family member is affected by the same RD as the individual or a different rare disease.\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Consent"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "n/a for Phenopackets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Disability"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "n/a for Phenopackets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the Data Model object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from phenopacket_mapper.data_standards import DataModel, DataField"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "rarelink_data_model = DataModel(\n",
    "    data_model_name=\"RareLink\",\n",
    "    resources=resources,  # Assuming resources are already defined\n",
    "    fields=[\n",
    "        # 1. Formal Criteria\n",
    "        # 1.1 Pseudonym\n",
    "        DataField(section=\"1. Formal Criteria\", ordinal=\"1.1\", name=\"Pseudonym\", value_set=vs_1_1),\n",
    "\n",
    "        # 1.2 Date of Admission\n",
    "        DataField(section=\"1. Formal Criteria\", ordinal=\"1.2\", name=\"Date of Admission\", value_set=vs_1_2),\n",
    "\n",
    "        # 2. Personal Information\n",
    "        # 2.1 Date of Birth\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.1\", name=\"Date of Birth\", value_set=vs_2_1),\n",
    "\n",
    "        # 2.2 Sex at Birth\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.2\", name=\"Sex at Birth\", value_set=vs_2_2),\n",
    "\n",
    "        # 2.3 Karyotypic Sex\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.3\", name=\"Karyotypic Sex\", value_set=vs_2_3),\n",
    "\n",
    "        # 2.4 Gender Identity\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.4\", name=\"Gender Identity\", value_set=vs_2_4),\n",
    "\n",
    "        # 2.5 Country of Birth\n",
    "        DataField(section=\"2. Personal Information\", ordinal=\"2.5\", name=\"Country of Birth\", value_set=vs_2_5),\n",
    "\n",
    "        # 3. Patient Status\n",
    "        # 3.1 Vital Status\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.1\", name=\"Vital Status\", value_set=vs_3_1),\n",
    "\n",
    "        # 3.2 Time of Death\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.2\", name=\"Time of Death\", value_set=vs_3_2),\n",
    "\n",
    "        # 3.3 Cause of Death\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.3\", name=\"Cause of Death\", value_set=vs_3_3),\n",
    "\n",
    "        # 3.4 Age Category\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.4\", name=\"Age Category\", value_set=vs_3_4),\n",
    "\n",
    "        # 3.5 Length of Gestation at Birth\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.5\", name=\"Length of Gestation at Birth\", value_set=vs_3_5),\n",
    "\n",
    "        # 3.6 Undiagnosed RD Case\n",
    "        DataField(section=\"3. Patient Status\", ordinal=\"3.6\", name=\"Undiagnosed RD Case\", value_set=vs_3_6),\n",
    "\n",
    "        # 4. Care Pathway\n",
    "        # 4.1 Encounter Start\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.1\", name=\"Encounter Start\", value_set=vs_4_1),\n",
    "\n",
    "        # 4.2 Encounter End\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.2\", name=\"Encounter End\", value_set=vs_4_2),\n",
    "\n",
    "        # 4.3 Encounter Status\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.3\", name=\"Encounter Status\", value_set=vs_4_3),\n",
    "\n",
    "        # 4.4 Encounter Class\n",
    "        DataField(section=\"4. Care Pathway\", ordinal=\"4.4\", name=\"Encounter Class\", value_set=vs_4_4),\n",
    "\n",
    "        # 5. Disease\n",
    "        # 5.1 Disease\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.1\", name=\"Disease\", value_set=vs_5_1_mondo),\n",
    "\n",
    "        # 5.2 Verification Status\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.2\", name=\"Verification Status\", value_set=vs_5_2),\n",
    "\n",
    "        # 5.3 Age at Onset\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.3\", name=\"Age at Onset\", value_set=vs_5_3),\n",
    "\n",
    "        # 5.4 Date of Onset\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.4\", name=\"Date of Onset\", value_set=vs_5_4),\n",
    "\n",
    "        # 5.5 Age at Diagnosis\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.5\", name=\"Age at Diagnosis\", value_set=vs_5_5),\n",
    "\n",
    "        # 5.6 Date of Diagnosis\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.6\", name=\"Date of Diagnosis\", value_set=vs_5_6),\n",
    "\n",
    "        # 5.7 Body Site [SNOMED CT]\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.7\", name=\"Body Site [SNOMED CT]\", value_set=vs_5_7),\n",
    "\n",
    "        # 5.8 Clinical Status\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.8\", name=\"Clinical Status\", value_set=vs_5_8),\n",
    "\n",
    "        # 5.9 Severity\n",
    "        DataField(section=\"5. Disease\", ordinal=\"5.9\", name=\"Severity\", value_set=vs_5_9),\n",
    "\n",
    "        # 6.1 Genetic Findings\n",
    "        # 6.1.1 Genomic Diagnosis\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.1\", name=\"Genomic Diagnosis [MONDO]\", value_set=vs_6_1_1),\n",
    "\n",
    "        # 6.1.2 Progress Status of Interpretation\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.2\", name=\"Progress Status of Interpretation\", value_set=vs_6_1_2),\n",
    "\n",
    "        # 6.1.3 Interpretation Status\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.3\", name=\"Interpretation Status\", value_set=vs_6_1_3),\n",
    "\n",
    "        # 6.1.4 Structural Variant Analysis Method\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.4\", name=\"Structural Variant Analysis Method\", value_set=vs_6_1_4),\n",
    "\n",
    "        # 6.1.5 Reference Genome\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.5\", name=\"Reference Genome\", value_set=vs_6_1_5),\n",
    "\n",
    "        # 6.1.6 Genetic Mutation String\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.6\", name=\"Genetic Mutation String\", value_set=vs_6_1_6),\n",
    "\n",
    "        # 6.1.7 Genomic DNA Change [g.HGVS]\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.7\", name=\"Genomic DNA Change [g.HGVS]\", value_set=vs_6_1_7),\n",
    "\n",
    "        # 6.1.8 Sequence DNA Change [c.HGVS]\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.8\", name=\"Sequence DNA Change [c.HGVS]\", value_set=vs_6_1_8),\n",
    "\n",
    "        # 6.1.9 Amino Acid Change [p.HGVS]\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.9\", name=\"Amino Acid Change [p.HGVS]\", value_set=vs_6_1_9),\n",
    "\n",
    "        # 6.1.10 Gene [HGNC-NR]\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.10\", name=\"Gene [HGNC-NR]\", value_set=vs_6_1_10),\n",
    "\n",
    "        # 6.1.11 Zygosity\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.11\", name=\"Zygosity\", value_set=vs_6_1_11),\n",
    "\n",
    "        # 6.1.12 Genomic Source Class\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.12\", name=\"Genomic Source Class\", value_set=vs_6_1_12),\n",
    "\n",
    "        # 6.1.13 DNA Change Type\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.13\", name=\"DNA Change Type\", value_set=vs_6_1_13),\n",
    "\n",
    "        # 6.1.14 Clinical Significance [ACMG]\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.14\", name=\"Clinical Significance [ACMG]\", value_set=vs_6_1_14),\n",
    "\n",
    "        # 6.1.15 Therapeutic Actionability\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.15\", name=\"Therapeutic Actionability\", value_set=vs_6_1_15),\n",
    "\n",
    "        # 6.1.16 Clinical Annotation Level Of Evidence\n",
    "        DataField(section=\"6.1 Genetic Findings\", ordinal=\"6.1.16\", name=\"Clinical Annotation Level Of Evidence\", value_set=vs_6_1_16),\n",
    "\n",
    "        # 6.2 Phenotypic Feature\n",
    "        # 6.2.1 Phenotypic Feature\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.1\", name=\"Phenotypic Feature\", value_set=vs_6_2_1),\n",
    "\n",
    "        # 6.2.2 Determination Date\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.2\", name=\"Determination Date\", value_set=vs_6_2_2),\n",
    "\n",
    "        # 6.2.3 Status\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.3\", name=\"Status\", value_set=vs_6_2_3),\n",
    "\n",
    "        # 6.2.4 Modifier\n",
    "        DataField(section=\"6.2 Phenotypic Feature\", ordinal=\"6.2.4\", name=\"Modifier\", value_set=vs_6_2_4),\n",
    "\n",
    "        # 6.3 Family History\n",
    "        # 6.3.0 Pseudonym\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.0\", name=\"Pseudonym\", value_set=vs_6_3_0),\n",
    "\n",
    "        # 6.3.1 Propositus/-a\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.1\", name=\"Propositus/-a\", value_set=vs_6_3_1),\n",
    "\n",
    "        # 6.3.2 Relationship of the individual to the index case / propositus/a\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.2\", name=\"Relationship of the individual to the index case / propositus/a\", value_set=vs_6_3_2),\n",
    "\n",
    "        # 6.3.3 Consanguinity\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.3\", name=\"Consanguinity\", value_set=vs_6_3_3),\n",
    "\n",
    "        # 6.3.4 Family Member Relationship\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.4\", name=\"Family Member Relationship\", value_set=vs_6_3_4),\n",
    "\n",
    "        # 6.3.5 Family Member Record Status\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.5\", name=\"Family Member Record Status\", value_set=vs_6_3_5),\n",
    "\n",
    "        # 6.3.6 Family Member Sex\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.6\", name=\"Family Member Sex\", value_set=vs_6_3_6),\n",
    "\n",
    "        # 6.3.7 Family Member Age\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.7\", name=\"Family Member Age\", value_set=vs_6_3_7),\n",
    "\n",
    "        # 6.3.8 Family Member Date of Birth\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.8\", name=\"Family Member Date of Birth\", value_set=vs_6_3_8),\n",
    "\n",
    "        # 6.3.9 Family Member Deceased\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.9\", name=\"Family Member Deceased\", value_set=vs_6_3_9),\n",
    "\n",
    "        # 6.3.10 Family Member Cause of Death [ICD10CM]\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.10\", name=\"Family Member Cause of Death [ICD10CM]\", value_set=vs_6_3_10),\n",
    "\n",
    "        # 6.3.11 Family Member Deceased Age\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.11\", name=\"Family Member Deceased Age\", value_set=vs_6_3_11),\n",
    "\n",
    "        # 6.3.12 Family Member Disease [MONDO]\n",
    "        DataField(section=\"6.3 Family History\", ordinal=\"6.3.12\", name=\"Family Member Disease [MONDO]\", value_set=vs_6_3_12),\n",
    "    ]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
